keps/sig-release/20190118-creating-packages.md

```diff
@@ -0,0 +1,417 @@
+[k-k-rpm]: https://github.com/kubernetes/kubernetes/tree/master/build/rpms
+
+To a good part this KEP tries to address most the [issue "k/k is the canonical source of all build artefacts"](https://github.com/kubernetes/kubernetes/issues/71677) and related
+issues:
+- [`kubeadm` package should depend on the same version of `kubelet` and
+  `kubectl` packages](https://github.com/kubernetes/kubernetes/issues/72871)
+- [Automate deb/rpm publishing](https://github.com/kubernetes/sig-release/issues/10)
+- ...
+
+
+
+### Goals
+
+- Fold the packaging specs, systemd unit files, dependency configuration, and
+  "Parts to reuse" into one place in `k/k`.
+
+  As a first step, the state from [`k/release/{rpm,debian}`][k-release] should
+  be used, as those are the things we release right now. As a second step all
+  the changes and additions from [`k/k/build/`][k-k] should be merged in.
+
+  There must be **one** package configuration per package, all different
+  packaging formats must use this as source of truth (a package configuration
+  could be *as an example* a config file for a package-generator plus
+  all the needed systemd unit files, ... )
+
+- The tooling must be used for cutting packages in the official release
+  process, it must also be usable in local development to generate one-off
+  packages.
+
+- <a id="pkg-list"></a>The tooling must support packaging
+  - `kubelet`
+  - `kubectl`
+  - `kubeadm`
+  - `cri-tools`
+  - `kubernetes-cni`
+
+  It should be easy to integrate new packages, especially when they are
+  one-binary packages
+
+- <a id="plat-list"></a>The tooling must support packaging for multiple platforms
+  - `linux/amd64`
+  - `linux/arm64`
+  - `linux/armhf`
+  - `linux/ppc64el`
+  - `linux/s390x`
+
+  It should be easy to integrate other platforms, given there is general
+  support by the supported distributions and the cross-building tools.
+  Examples: `kubectl` for `darwin/amd64`, `kubelet` for `windows/386`, ...
+
+- The tooling must support building `deb`s and `rpm`s
+
+  It should be possible to add new package types, e.g. NuGet for Windows
+
+- The tooling must support signing the generated packages
+
+- The documentation for the process of building packages, local one-off builds
+  and as part of the release process, must be excellent.
+
+- Any other option to build packages must be retired eventually
+
+  There should be documentation how and where the old packaging option can be
+  found and how they can be used for cases where someone wants to build
+  packages before the new tools were introduced.
+
+- It must be equally easy to build packages from
+  - the currently checked out revision (any revision after the tooling has been
+    introduced into `k/k`)
+  - a pre-build binary (pulling from some well-defined staging bucket or that like)
+
+- It must be containerized and must not depend on anything more then docker on
+  the host
+
+### Non-Goals
+
+- Uploading, publishing, ... of packages
```

Comment:
yet, that's one of the bigger problems ATM:
- only folks from google can do that.
- we don't have a repo for pre-releases.

we might want to have uploads as part of the tooling in k/k, so that anyone with credentials can do it.
such credentials can potentially rotate on release.

Reply:
@dims had a suggestion for creating a separate repo that is outside Google. I realize this is not the goal for this KEP. just wanted to bring this to attention of @dims so that we can tackle this in a separate KEP.

Reply:
Yes, there needs to be another KEP (or folded into this KEP?). And there are discussions going on about publishing. I started somewhere and I am planning to create another KEP for the publishing part, but I am also happy if someone wants to take that part over.

And yes: I am pretty sure we want multiple repos for stable & pre-release or such.